edgePy/DGEList.py

```diff
@@ -292,17 +292,40 @@ def rpkm(
         # self = self.cpm(log=log, prior_count=prior_count)
 
     def tpm(
-        self, transcripts: Mapping, log: bool = False, prior_count: float = PRIOR_COUNT
-    ) -> None:
-        """Return the DGEList normalized to reads per kilobase of transcript
-        length.
+        self,
+        gene_lengths: np.ndarray,
+        log: bool = False,
+        prior_count: float = PRIOR_COUNT,
+        mean_fragment_lengths: np.ndarray = None,
+    ) -> np.ndarray:
+        """Normalize the DGEList to transcripts per million
+
+        Args:
+            gene_lengths: vector of length for each gene in counts table
+            log: store log outputs
+            prior_count:
+            mean_fragment_lengths: vector of mean fragment length for each sample in counts table (optional)
 
         """
-        raise NotImplementedError
 
-        # TODO: Implement here
+        # based on https://haroldpimentel.wordpress.com/2014/05/08/what-the-fpkm-a-review-rna-seq-expression-units/ and
+        # https://gist.github.com/slowkow/c6ab0348747f86e2748b
+        # how many counts per base
 
-        # self = self.cpm(log=log, prior_count=prior_count)
+        # compute effective length not allowing negative lengths
+        if mean_fragment_lengths:
+            effective_lengths = (gene_lengths[:, None] - mean_fragment_lengths[None, :]).clip(
```

Comment:
What do you think about this version for readability sake?

```python
effective_lengths = np.subtract(gene_lengths, mean_fragment_lengths).clip(min=1)
```

Or:

```python
effective_lengths = (gene_lengths - mean_fragment_lengths).clip(min=1)
```

Why did you need to slice?

Reply:
One is an array of shape (n,) the other one of shape (m,), the result needs to be a matrix of shape (n,m) as effective lengths depends on both sample and gene. 

I'll change for np.newaxis as it is more explicit and people directly know what to look/google for if they are not comfortable with numpy (https://stackoverflow.com/questions/944863/numpy-should-i-use-newaxis-or-none)

Reply:
Thanks! I had no idea one array was a row, and the other a column. A bit o' docs (in the args description) and an assertion of `np.array.shape()` will help.

I understand the need for slicing now! Would it be as cheap to transpose the second array?

```python
effective_lengths = (gene_lengths - mean_fragment_lengths.T).clip(min=1)
```

Reply:
I think that would not work. You subtract an array of dimension (n,) minus an array dimension (m, 1). You need the np.newaxis (or None)